internal/kafka/internal/services/kafka.go

```diff
@@ -960,52 +962,107 @@ type KafkaStatusCount struct {
+		return nil, apiErrors.NewWithCause(apiErrors.ErrorGeneral, err, "failed to list clusters")
+	}
 
-	var kafkas []*dbapi.KafkaRequest
+	streamingUnitsCountPerRegion := []KafkaStreamingUnitCountPerRegion{}
+
+	// pre-populate regions count with zero count values.
+	// This is useful and it ensures that the count drops to 0 for each instance type and region when all the kafkas are removed
+	for _, cluster := range clusters {
+		supportedInstanceTypes := strings.Split(cluster.SupportedInstanceType, ",")
 
+		for _, supportedInstanceType := range supportedInstanceTypes {
+			instanceType := strings.TrimSpace(supportedInstanceType)
+			if instanceType == "" {
+				continue
+			}
+
+			streamingUnitsCountPerRegion = append(streamingUnitsCountPerRegion, KafkaStreamingUnitCountPerRegion{
+				CloudProvider: cluster.CloudProvider,
+				ClusterId:     cluster.ClusterID,
+				InstanceType:  instanceType,
+				Region:        cluster.Region,
+				Count:         0,
+			})
+		}
+	}
+
+	dbConn = k.connectionFactory.New()
+	var kafkasPerRegion []*KafkaPerRegionCount
 	if err := dbConn.Model(&dbapi.KafkaRequest{}).
-		Scan(&kafkas).Error; err != nil {
+		Select("cloud_provider, region, count(1) as Count, size_id, cluster_id, instance_type").
+		Group("size_id, cluster_id, cloud_provider, region, instance_type").
+		Scan(&kafkasPerRegion).Error; err != nil {
 		return nil, errors.NewWithCause(errors.ErrorGeneral, err, "Failed to count kafkas when setting capacity metrics")
 	}
 
-	var results []KafkaRegionCount
-
-	for _, kafka := range kafkas {
+	for _, kafkaCountPerRegion := range kafkasPerRegion {
 		resultPresent := false
-		instSize, err := k.kafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
+		instSize, err := k.kafkaConfig.GetKafkaInstanceSize(kafkaCountPerRegion.InstanceType, kafkaCountPerRegion.SizeId)
 		if err != nil {
-			return nil, errors.NewWithCause(errors.ErrorGeneral, err, "Failed to count kafkas of '%s' instance type and '%s' size_id when setting capacity metrics", kafka.InstanceType, kafka.SizeId)
+			return nil, errors.NewWithCause(errors.ErrorGeneral, err, "Failed to count kafkas of '%s' instance type and '%s' size_id when setting capacity metrics", kafkaCountPerRegion.InstanceType, kafkaCountPerRegion.SizeId)
 		}
 
-		for i, result := range results {
-			if result.CloudProvider == kafka.CloudProvider && result.ClusterId == kafka.ClusterID &&
-				result.InstanceType == kafka.InstanceType && result.Region == kafka.Region {
-				results[i].Count = result.Count + float64(instSize.CapacityConsumed)
+		streamingUnitCount := float64(instSize.CapacityConsumed) * kafkaCountPerRegion.Count
+		for i, streamingUnitCountPerRegion := range streamingUnitsCountPerRegion {
+			if streamingUnitCountPerRegion.isSame(kafkaCountPerRegion) {
+				streamingUnitsCountPerRegion[i].Count += streamingUnitCount
 				resultPresent = true
+				break
 			}
 		}
+
 		if !resultPresent {
```

Comment:
This might be useful when a kafka is not assigned to a cluster yet (in the dynamic scaling mode).
The metric will be further revisited then as well if needed for now I left the condition as it was before.

Reply:
Can we add a comment about that here just as a reminder :)